PowerEditor/src/WinControls/OpenSaveFileDialog/FileDialog.cpp

```diff
@@ -267,6 +267,7 @@ TCHAR * FileDialog::doSaveDlg()
 
 	NppParameters& params = NppParameters::getInstance();
 	_ofn.lpstrInitialDir = params.getWorkingDir();
+	_ofn.lpstrDefExt = _defExt.c_str();

```

Comment:
**Very interesting this line**: It's fixed the problem that auto-adding file ext (by changing file type) does not work in new style dialog.
I was searching to fix this problem but the only solution I found was using ugly COM Object.
I still don't get why and how this line is the solution so @Uhf7 if you have any explanation please let me know.

Reply:
It's actually a rather complex mechanism of **GetOpenFileName** and **GetSaveFileName**.

If the field `_ofn.lpstrDefExt = NULL`, then there is no default extension specified, and if there is no default extension, then no extension is appended automatically ever.

If the field `_ofn.lpstrDefExt` points to an empty string (`_ofn.lpstrDefExt = L""`) , the behavior changes: Now, the selected extension from the _file types drop down_ is automatically appended. And, if there is nothing selected in the _file types drop down_ (`"*.*"`), then `_ofn.lpstrDefExt` is appended, which is in this example nothing.

If the field `_ofn.lpstrDefExt` points to a not-empty string (`_ofn.lpstrDefExt = L"npj"`), then the selected extension from the _file types drop down_ is automatically appended, regardless of the `"npj"` in `_ofn.lpstrDefExt`. Only if nothing is selected in the _file types drop down_ (`"*.*"`), then the default ending is appended, which would then be `".npj"`.

A point to be watched carefully is the handling of the starting dot of the default extension. If there is an empty string specified as default extension, then no dot is added. If there is a not-empty string specified as default extension, then the dot is inserted automatically. If the specified default extension starts with a dot already, then we have a double dots. Means for this project: We could enforce a no-dot policy for the default session and workspace filenames in _Preferences-->MISC_.

Another point I realized while writing this is, that I forgot to improve the load procedures for workspace and session files too. The line `_ofn.lpstrDefExt = _defExt.c_str();` should also be included into `FileDialog::doOpenSingleFileDlg();`. Then the **Open Workspace** function works automatically, the **Open Session** function needs a `fDlg.setDefExt(ext)` call too at loading. Could force-push a new Version of this PR easily with this modifications.

Reply:
@Uhf7 Thank you for the very detailed explanation.

> that I forgot to improve the load procedures for workspace and session files too. 

Ah, too late. Creating a new PR will be fine. Thank you again for fixing this bug.